pennylane_lightning_gpu/lightning_gpu.py

```diff
@@ -786,32 +788,42 @@ def expval(self, observable, shot_range=None, bin_size=None):
                 return np.squeeze(np.mean(samples, axis=0))
 
             if observable.name in ["SparseHamiltonian"]:
-                CSR_SparseHamiltonian = observable.sparse_matrix().tocsr()
-                return self._gpu_state.ExpectationValue(
-                    CSR_SparseHamiltonian.indptr,
-                    CSR_SparseHamiltonian.indices,
-                    CSR_SparseHamiltonian.data,
-                )
+                if not self._mpi:
+                    CSR_SparseHamiltonian = observable.sparse_matrix().tocsr()
+                    return self._gpu_state.ExpectationValue(
+                        CSR_SparseHamiltonian.indptr,
+                        CSR_SparseHamiltonian.indices,
+                        CSR_SparseHamiltonian.data,
+                    )
+                else:
+                    raise RuntimeError(
+                        "LightningGPU-MPI does not currently support SparseHamiltonian."
+                    )
 
             if observable.name in ["Hamiltonian"]:
-                device_wires = self.map_wires(observable.wires)
-                # 16 bytes * (2^13)^2 -> 1GB Hamiltonian limit for GPU transfer before
-                if len(device_wires) > 13:
-                    coeffs = observable.coeffs
-                    pauli_words = []
-                    word_wires = []
-                    for word in observable.ops:
-                        compressed_word = []
-                        if isinstance(word.name, list):
-                            for char in word.name:
-                                compressed_word.append(_name_map[char])
-                        else:
-                            compressed_word.append(_name_map[word.name])
-                        word_wires.append(word.wires.tolist())
-                        pauli_words.append("".join(compressed_word))
-                    return self._gpu_state.ExpectationValue(pauli_words, word_wires, coeffs)
-
+                if not self._mpi:
```

Comment:
Do we need to branch here? Can this path serve both MPI and non MPI?

Reply:
Not really, with single GPU backend, this Pauli word based expval is based on `custatevecComputeExpectationsOnPauliBasis`, which accept an array of Pauli word.

Reply:
So we will need to map the global indices to local ones, then convert the wires on the words to match the mapped wires to support this ? If so, let's visit this later with that plan in mind.

Reply:
Correct! we will need to map the global indices to local ones and a potential failure will happen if there are `obs` at each wires and this means there will be no wires available for the indices mapping.

Reply:
fix in PR#119.